sdk/messaging/azservicebus/processor.go

```diff
@@ -189,47 +181,64 @@ func newProcessor(ns legacyNamespace, options ...ProcessorOption) (*Processor, e
 // sent to your handleError function. The processor will retry and restart as needed -
 // no user intervention is required.
 func (p *Processor) Start(handleMessage func(message *ReceivedMessage) error, handleError func(err error)) error {
-	select {
-	case <-p.Done():
-		return errClosed{link: "processor"}
-	default:
-	}
-
 	p.mu.Lock()
 	defer p.mu.Unlock()
 
-	p.activeReceiversWg.Add(1)
+	// TODO: stop multiple invocations!
+	/*
+		p.handleMessage = ()
+		p.handleError = func() {
+			tab.For(ctx).Debug()
+		}
+	*/
+
+	p.wg.Add(1)
+
 	go func(ctx context.Context) {
-		defer p.activeReceiversWg.Done()
+		defer func() {
+			p.wg.Done()
+			tab.For(ctx).Debug("Exiting forever loop in Processor")
+		}()
 
 		for {
 			retry, _ := amqpCommon.Retry(p.config.RetryOptions.Times, p.config.RetryOptions.Delay, func() (interface{}, error) {
-				receiver, err := p.ns.NewReceiver(ctx, p.config.FullEntityPath,
-					internal.ReceiverWithReceiveMode(internal.ReceiveMode(p.config.ReceiveMode)))
+				_, receiver, _, _, err := p.amqpLinks.Get(ctx)
 
 				if err != nil {
+					if !isRetryableLinksError(err) {
+						tab.For(ctx).Debug(fmt.Sprintf("Non retriable error occurred ('%s'), stopping the processor loop", err.Error()))
+						handleError(fmt.Errorf("Non-retryable error, stopping processor loop: %w", err))
+						return false, nil
+					}
+
 					// notify the user and then fall into doing a retry
-					handleError(err)
+					handleError(fmt.Errorf("failed getting links for subscribe, will retry: %w", err))
 					return true, amqpCommon.Retryable("")
 				}
 
-				defer receiver.Close(ctx)
-
 				if err := receiver.IssueCredit(uint32(p.config.MaxConcurrentCalls)); err != nil {
 					// notify the user but there's no reason to restart because this failure must be
 					// an internal error.
-					handleError(err)
-					return false, nil
+					handleError(fmt.Errorf("internal failure when issuing credit, will recreate links: %w", err))
+					return true, nil
 				}
 
 				// we retry infinitely, but do it in the pattern they specify via their retryOptions for each "round" of retries.
-				retry := p.subscribe(ctx, receiver, p.config.ShouldAutoComplete, handleMessage, handleError)
+				err = p.subscribe(ctx, receiver, handleMessage, handleError)
 
-				if retry {
-					return true, amqpCommon.Retryable("")
+				if err != nil {
+					if isRetryableSubscribeError(err) {
```

Comment:
I believe this line can be condensed into its parent?

Reply:
I don't think so, only because there is a default 'err != nil' behavior too (ie, if it's not retryable we exit the retries altogether).

TBH, there are some issues with how errors are handled here that I am planning on going through in a recovery PR. I've learned quite a bit since I first wrote this. :)

Reply:
Ah you're right, I missed that.